[FIRRTL] Handle flips in const connections

trilorez opened the pull request:
Addresses #5209 

I can also update `areTypesWeaklyEquivalent`, although `const` types were introduced into the firrtl spec in the same version that removed partial connection, so I suppose I can leave it. Or I can remove all of the const logic it has to clean things up.

youngar commented:
I think we still accept partial connects, its probably best to update `areTypesWeaklyEquivalent` until all callers are fully removed.

youngar commented on test/Dialect/FIRRTL/connect-errors.mlir:
```diff
@@ -551,11 +551,48 @@ firrtl.module @test(in %a : !firrtl.uint<1>, out %b : !firrtl.sint<1>) {
+
+// -----
+
+/// Non-const doube flip types cannot be connected to const.
```
`s/doube/double`

youngar commented on lib/Dialect/FIRRTL/FIRRTLTypes.cpp:
```diff
@@ -690,28 +690,41 @@ std::pair<uint64_t, bool> FIRRTLBaseType::rootChildFieldID(uint64_t fieldID,
 /// definition of type equivalence in the FIRRTL spec.  If the types being
 /// compared have any outer flips that encode FIRRTL module directions (input or
 /// output), these should be stripped before using this method.
 bool firrtl::areTypesEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
+                                bool destOuterTypeIsConst,
                                 bool srcOuterTypeIsConst,
                                 bool requireSameWidths) {
 
-  // If the types are trivially equal, return true.
-  if (destFType == srcFType)
-    return true;
+  // If the types are trivially equal, return true as long as outer constness is
+  // compatible.
+  if (destFType == srcFType) {
+    if (destOuterTypeIsConst == srcOuterTypeIsConst)
+      return true;
+    if (destFType.isGround())
+      return !destOuterTypeIsConst || srcOuterTypeIsConst;
+  }
```
Hmm maybe we should just get rid of the early break.  I think I added this recently, and introduced the bug of ignoring the outer const type.  If we ever push const to the leaves or something like that, then maybe we can introduce the early break again, wdyt?

Also I suggest we add a test case for this because I keep thinking we can add this trivial early break.  Its now covered by the flip test cases you added, but the most basic test might be:
```mlir
/// Const types cannot be connected to non-const types.

firrtl.circuit "test" {
firrtl.module @test(in %in   : !firrtl.const.bundle<a: uint<1>>,
                    out %out : !firrtl.bundle<a: uint<1>>) {
  // expected-error @+1 {{type mismatch}}
  firrtl.connect %out, %in : !firrtl.bundle<a: uint<1>>, !firrtl.const.bundle<a: uint<1>>
}
}
```

trilorez replied:
Yeah, the outer constness gets kind of unwieldy, I like your suggestion of keeping things in one place for ground types.
Removed, and added test case (I assume you accidentally wrote the valid version).

youngar commented on lib/Dialect/FIRRTL/FIRRTLTypes.cpp:
```diff
@@ -720,8 +733,11 @@ bool firrtl::areTypesEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
+  bool srcIsConst = srcOuterTypeIsConst || srcFType.isConst();
+  bool destIsConst = destOuterTypeIsConst || destFType.isConst();
+
+  // A 'const' ground destination requires a 'const' source.
+  if (destType.isGround() && destIsConst && !srcIsConst)
     return false;
```
Maybe we should keep all the ground-related logic grouped at the end?  The current logic down below sort of looks wrong to me if this were to fall through:
```c++
  // Ground types can be connected if they are the same, or the source type is
  // a const version of the destination type.
  return destType == srcType || destType == srcType.getConstType(false);
```
Should this be something like:
```c++
if (destIsConst && !srcIsConst)
    return false;
return destType.getConst(false) == srcType.getConst(false);

// OR:

auto realDestType = destType.getConstType(destIsConst);
return realDestType == srcType.getConstType(true) ||  realDestType == srcType.getConstType(false);
```

trilorez replied:
I grouped the ground stuff at the end, and I also moved the reset-type tests below the aggregates as well.

trilorez commented:
I updated `areTypesWeaklyEquivalent`. Const parser support isn't in yet, I can add const partial connect tests on the parsing pr.
I also noticed an edge case where flips weren't being taken into account for `ResetType`, so I fixed that and added a test.

youngar reviewed (approved):
🥳
